python-checks/src/main/java/org/sonar/python/checks/UnresolvedSymbolsCheck.java

```diff
@@ -0,0 +1,94 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.FileInput;
+import org.sonar.plugins.python.api.tree.ImportFrom;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.Tree;
+
+@Rule(key = "S5953")
+public class UnresolvedSymbolsCheck extends PythonSubscriptionCheck {
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FILE_INPUT, ctx -> {
+      FileInput fileInput = (FileInput) ctx.syntaxNode();
+      if (importsManipulatedAllProperty(fileInput)) {
+        return;
+      }
+      UnresolvedSymbolsVisitor unresolvedSymbolsVisitor = new UnresolvedSymbolsVisitor();
+      fileInput.accept(unresolvedSymbolsVisitor);
+      if (!unresolvedSymbolsVisitor.callGlobalsOrLocals && !unresolvedSymbolsVisitor.hasUnresolvedWildcardImport) {
+        addNameIssues(unresolvedSymbolsVisitor.nameIssues, ctx);
+      }
+    });
+  }
+
+  private static boolean importsManipulatedAllProperty(FileInput fileInput) {
+    return fileInput.globalVariables().stream().anyMatch(s -> s.name().equals("__all__") && s.fullyQualifiedName() != null);
+  }
+
+  private static void addNameIssues(Map<String, List<Name>> nameIssues, SubscriptionContext subscriptionContext) {
+    nameIssues.forEach((name, list) -> {
+      Name first = list.get(0);
+      PreciseIssue issue = subscriptionContext.addIssue(first, first.name() + " is not defined. Change its name or define it before using it");
+      list.stream().skip(1).forEach(n -> issue.secondary(n, null));
+    });
+  }
+
+  private static class UnresolvedSymbolsVisitor extends BaseTreeVisitor {
+
+    private boolean hasUnresolvedWildcardImport = false;
+    private boolean callGlobalsOrLocals = false;
+    private final Map<String, List<Name>> nameIssues = new HashMap<>();
+
+    @Override
+    public void visitName(Name name) {
+      if (name.isVariable() && name.symbol() == null) {
+        nameIssues.computeIfAbsent(name.name(), k -> new ArrayList<>()).add(name);
+      }
+    }
+
+    @Override
+    public void visitImportFrom(ImportFrom importFrom) {
+      hasUnresolvedWildcardImport |= importFrom.hasUnresolvedWildcardImport();
```

Comment:
Maybe we can consider avoiding raising issues on files which use any wildcard imports. Even though the rule is going out of Sonar_way, I feel our resolution of such import is not strong enough to raise issues here (if `moduleA` wildcard imports `moduleB` and then `moduleC` wildcard imports `moduleA`, we will fail to resolve symbols from `moduleB` despite them being available to use (and `hasUnresolvedWildcardImport()` won't necessarily be `true` if I'm not mistaken).
 
Anyway, if you agree with this, I believe it can be done in a separate ticket.

Reply:
Agreed, Let's handle this in a separate ticket